doc: mgmt: mcumgr: Add list of libraries/tools

nordicjm opened the pull request:
Adds a list of tools and libraries that are available

nordicjm committed:
doc: mgmt: mcumgr: Add list of libraries/tools

Adds a list of tools and libraries that are available

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
@kartben do we have better tick/cross symbols to use in doc pages?

kartben commented:
> @kartben do we have better tick/cross symbols to use in doc pages?

✓ / ✕ or ✔ / ✖, please. 
A bit concerned about the super wide table (definitely broken on the PDF output) so if you can find smart ways to present the information differently, that would be great. I wouldn't consider this blocking though, and worst case scenario we could even exclude the section from the PDF. One thing that comes to mind would be removing iOS and Android column for now (still far too wide, but it helps :)).

kartben commented on doc/services/device_mgmt/mcumgr.rst:
while it's definitely not possible and completely ok for the table to be >100 char, other doc content should wrap at 100.

kartben commented:
> > @kartben do we have better tick/cross symbols to use in doc pages?
> 
> ✓ / ✕ or ✔ / ✖, please. 

dark theme issue with the later, eh? Unicode is so much fun.

nordicjm commented:
> > > @kartben do we have better tick/cross symbols to use in doc pages?
> > 
> > 
> > ✓ / ✕ or ✔ / ✖, please.
> 
> dark theme issue with the later, eh? Unicode is so much fun.

Isn't it just! :smile:

kartben reviewed (commented):
So this basically looks good to me (see one comment re: in-tree client), I am simply not sure what to do with the PDF. Since I don't think that even with the best typesetting engine we could find this table would ever look good or be automatically reformatted to fit in A4 constraints, I think we are looking at either removing the table from the PDF altogether (`.. only:: html`), or there should be another variation of this information that suits a page output better (in a `.. only:: latex` block). For the latter, it wouldn't be terrible to just have a bullet list with all the various links, and maybe adding "(license, app type)" ..?

I know this is extra work but, again, I don't think there can be a magic solution in this particular example.

nordicjm commented:
@kartben is there a way it can be forced to be rotated for the PDF so it is landscape instead of portrait? Then it should fit I would think

nordicjm commented on doc/services/device_mgmt/mcumgr.rst:
```diff
@@ -33,6 +33,42 @@ the Zephyr tree.
+    | and `iOS                                                                       |         |       |     |        |          |        |           |     |    |     |      |          |    |       |        |               | Swift      |         |
+    | <https://github.com/NordicSemiconductor/IOS-nRF-Connect-Device-Manager>`_      |         |       |     |        |          |        |           |     |    |     |      |          |    |       |        |               |            |         |
+    +--------------------------------------------------------------------------------+---------+-------+-----+--------+----------+--------+-----------+-----+----+-----+------+----------+----+-------+--------+---------------+------------+---------+
+    | Zephyr MCUmgr client (in-tree)                                                 | ✕       | ✓     | ✕   | ✕      | ✓        | ✓      | ✓         | ✓   | ✓  | ✓   | ✕    | ✕        | ✕  | ✕     | ✕      | Library       | C          | Apache  |
```
Unfortunately not, I thought I asked the submitter to add documentation but seemingly I didn't, so no docs, just the files and unit tests

kartben replied:
then the below would probably be quite helpful to make it easier to locate it. Full table to save you the hassle

```

    +---------------------------------------------------------------------------------------------+-------------------------------------------+--------------------------+--------------------------------------------------+---------------+------------+---------+
    | Name                                                                                        | OS support                                | Transports               | Groups                                           | Type          | Language   | License |
    |                                                                                             +---------+-------+-----+--------+----------+--------+-----------+-----+----+-----+------+----------+----+-------+--------+               |            |         |
    |                                                                                             | Windows | Linux | mac | Mobile | Embedded | Serial | Bluetooth | UDP | OS | IMG | Stat | Settings | FS | Shell | Zephyr |               |            |         |
    +=============================================================================================+=========+=======+=====+========+==========+========+===========+=====+====+=====+======+==========+====+=======+========+===============+============+=========+
    | `AuTerm <https://github.com/thedjnK/AuTerm/>`_                                              | ✓       | ✓     | ✓   | ✕      | ✕        | ✓      | ✓         | ✓   | ✓  | ✓   | ✓    | ✓        | ✓  | ✓     | ✓      | Application   | C++ (Qt)   | GPLv3   |
    +---------------------------------------------------------------------------------------------+---------+-------+-----+--------+----------+--------+-----------+-----+----+-----+------+----------+----+-------+--------+---------------+------------+---------+
    | `mcumgr-client <https://github.com/vouch-opensource/mcumgr-client/>`_                       | ✓       | ✓     | ✓   | ✕      | ✕        | ✓      | ✕         | ✕   | ✕  | ✓   | ✕    | ✕        | ✕  | ✕     | ✕      | Application   | Rust       | BSD     |
    +---------------------------------------------------------------------------------------------+---------+-------+-----+--------+----------+--------+-----------+-----+----+-----+------+----------+----+-------+--------+---------------+------------+---------+
    | `mcumgr-web <https://github.com/boogie/mcumgr-web/>`_                                       | ✓       | ✓     | ✓   | ✕      | ✕        | ✕      | ✓         | ✕   | ✕  | ✓   | ✕    | ✕        | ✕  | ✕     | ✕      | Web page      | Javascript | MIT     |
    |                                                                                             |         |       |     |        |          |        |           |     |    |     |      |          |    |       |        | (chrome only) |            |         |
    +---------------------------------------------------------------------------------------------+---------+-------+-----+--------+----------+--------+-----------+-----+----+-----+------+----------+----+-------+--------+---------------+------------+---------+
    | nRF Connect Device Manager: |br|                                                            |         |       |     |        |          |        |           |     |    |     |      |          |    |       |        |               |            |         |
    | `Android                                                                                    | ✕       | ✕     | ✕   | ✓      | ✕        | ✕      | ✓         | ✕   | ✓  | ✓   | ✓    | ✓        | ✓  | ✓     | ✓      | Library and   | Java,      | Apache  |
    | <https://github.com/NordicSemiconductor/Android-nRF-Connect-Device-Manager/>`_              |         |       |     |        |          |        |           |     |    |     |      |          |    |       |        | application   | Kotlin,    |         |
    | and `iOS                                                                                    |         |       |     |        |          |        |           |     |    |     |      |          |    |       |        |               | Swift      |         |
    | <https://github.com/NordicSemiconductor/IOS-nRF-Connect-Device-Manager>`_                   |         |       |     |        |          |        |           |     |    |     |      |          |    |       |        |               |            |         |
    +---------------------------------------------------------------------------------------------+---------+-------+-----+--------+----------+--------+-----------+-----+----+-----+------+----------+----+-------+--------+---------------+------------+---------+
    | Zephyr MCUmgr client (:zephyr_file:`in-tree <subsys/mgmt/mcumgr/smp_client/src/client.c>`)  | ✕       | ✓     | ✕   | ✕      | ✓        | ✓      | ✓         | ✓   | ✓  | ✓   | ✕    | ✕        | ✕  | ✕     | ✕      | Library       | C          | Apache  |
    +---------------------------------------------------------------------------------------------+---------+-------+-----+--------+----------+--------+-----------+-----+----+-----+------+----------+----+-------+--------+---------------+------------+---------+
```

nordicjm replied:
That file is not overly useful, that's just SMP, there's also groups (with separate Kconfigs). MCUmgr client support really needs to be documented @jheiskan81

kartben replied:
kkk

kartben commented:
> @kartben is there a way it can be forced to be rotated for the PDF so it is landscape instead of portrait? Then it should fit I would think
> 
> Had a look and found this https://stackoverflow.com/questions/50620759/sphinx-doc-generate-a-landscape-page-for-long-table-or-graph so seems possible with an extension?

Ya I was about to point to the exact same thread! Seems pretty straightforward, let me give it a try (assuming you might not want to test building the PDF locally :))

kartben commented:
@nordicjm  close enough but... meh

<img width="1018" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/128251/5b16ceb4-39eb-4192-9246-320e5794a2cf">

nordicjm commented:
That is very disappointing

nordicjm commented:
@kartben see if this helps with the pdf version

de-nordic commented:
Shouldn't there also be the mcumgr-cli GO(ne decrepit) client?

nordicjm commented:
> Shouldn't there also be the mcumgr-cli GO(ne decrepit) client?

That would need it's own category:
![User-trash-full](https://github.com/zephyrproject-rtos/zephyr/assets/40387179/2084b655-fb0b-46e5-af95-51c566892dff)

de-nordic commented:
Maybe the array should be divided into two:
1) OS supported, transport , license
2) groups

kartben committed:
doc: latex: enable pdflscape package

Enable the pdflscape package (available in the Latex distros recommended
 for Zephyr) to allow for some pages to be rendered in landscape mode.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben commented:
@nordicjm I didn't want to push force to your branch as I've add bad experiences with it before, but see the two added commits, you'll want to make the pdflscape one first as part of your rebase. Thanks!

There are still some stupid layout decisions made by LaTeX it would seem, but this is good enough I'd say.

<img width="1071" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/128251/96104beb-1fe7-4d32-b618-cd275f6ab724">

nordicjm commented:
Made some additional minifications to put things on multiple lines, hopefully this works slightly better

kartben commented:
> Made some additional minifications to put things on multiple lines, hopefully this works slightly better

I'll check but I don't think `br` does anything in LaTeX output (probably should... I will try to look into it), hence why I only left it at adding white spaces to the X/Y/Z "words" to try and have them break a bit more easily

nordicjm commented:
@kartben did the break work in the PDF?

nordicjm commented:
Let's go with it as-is, and see if we can maybe improve it at a later date
